components/nimbus/src/matcher.rs

```diff
@@ -54,3 +56,35 @@ pub struct AppContext {
     #[serde(flatten)]
     pub custom_targeting_attributes: Option<Map<String, Value>>,
 }
+
+/// The `AppContext` object represents the parameters and characteristics of the
+/// consuming application that we are interested in for targeting purposes. The
+/// `app_name`, `app_id` and `channel` fields are not optional as they are expected
+/// to be provided by all consuming applications as they are used in the top-level
+/// targeting that help to ensure that an experiment is only processed
+/// by the correct application.
+///
+/// Definitions of the fields are as follows:
+/// - `app_name`: This is the name of the application (e.g. "Fenix" or "Firefox iOS")
+/// - `app_id`: This is the application identifier, especially for mobile (e.g. "org.mozilla.fenix")
+/// - `channel`: This is the delivery channel of the application (e.g "nightly")
+/// - `app_version`: The user visible version string (e.g. "1.0.3")
+/// - `app_build`: The build identifier generated by the CI system (e.g. "1234/A")
+/// - `locale`: The locale of the application during initialization (e.g. "es-ES")
+/// - `os`: The name of the operating system (e.g. "Android", "iOS", "Darwin", "Windows")
+/// - `os_version`: The user-visible version of the operating system (e.g. "1.2.3")
+/// - `custom_targeting_attributes`: Contains attributes specific to the application, derived by the application
+#[cfg(not(feature = "stateful"))]
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+pub struct AppContext {
+    pub app_name: String,
+    pub app_id: String,
+    pub channel: String,
+    pub app_version: Option<String>,
+    pub app_build: Option<String>,
+    pub locale: Option<String>,
+    pub os: Option<String>,
+    pub os_version: Option<String>,
+    #[serde(flatten)]
+    pub custom_targeting_attributes: Option<Map<String, Value>>,
+}
```

Comment:
At some point, I prefer this re-declaration approach over the preffing on an off of properties.

I think you have it right here.

Reply:
Same, it was so many `#[cfg(feature = "")]`'s one after the other and I couldn't wrap it in a multiline because `cfg_if` doesn't seem to support wrapping around struct members.